File: Server/PluginSDK/TS/src/runner.ts
/** 将 Node 标准输入的 Sidecar JSONL 映射到普通插件函数。 */

import { createInterface } from "node:readline";
import { stdin, stdout } from "node:process";
import { pathToFileURL } from "node:url";
import { resolve } from "node:path";
import { Plugin } from "./plugin.js";
import type { RuntimeInvocation } from "./model.js";

type ProtocolMessage = Record<string, unknown>;

export interface RunnerOptions { readonly concurrentInvocations?: boolean; }

class JsonLineWriter {
  private tail = Promise.resolve();

  /** 把输出追加到单一 Promise 队列，保证并发结果不会交错成坏 JSON。 */
  public write(message: ProtocolMessage): Promise<void> {
    const encoded = `${JSON.stringify(message)}\n`;
    const operation = this.tail.then(() => new Promise<void>((resolveWrite, rejectWrite) => {
      stdout.write(encoded, (error) => error ? rejectWrite(error) : resolveWrite());
    }));
    this.tail = operation.catch(() => undefined);
    return operation;
  }
}

/** 严格执行 initialize→invoke*→stop；每个请求以 requestId 精确关联。 */
export async function serve(plugin: Plugin, options: RunnerOptions = {}): Promise<void> {
  let initialized = false;
  let stopped = false;
  const writer = new JsonLineWriter();
  const activeTasks = new Set<Promise<void>>();

  /** 校验当前 Host 从 1 递增的 u64 请求号；超出 JS 安全整数时拒绝而不改写精度。 */
  function parseRequestId(message: ProtocolMessage): number {
    const requestId = message.requestId;
    if (typeof requestId !== "number" || !Number.isSafeInteger(requestId) || requestId < 0) {
      throw new Error("requestId 必须是非负 JavaScript 安全整数")
    }
    return requestId;
  }

  /** 执行一个作者函数并回显请求身份；所有插件异常转换为 error 帧。 */
  async function invoke(message: ProtocolMessage): Promise<void> {
    const requestId = parseRequestId(message);
    try {
      const action = await plugin.invoke(message.invocation as RuntimeInvocation);
      await writer.write({ type: "result", requestId, action });
    } catch (error) {
      await writer.write({ type: "error", requestId, message: error instanceof Error ? `${error.name}: ${error.message}` : String(error) });
    }
  }

  /** 等待所有在途作者任务，再且仅执行一次停止生命周期。 */
  async function waitAndStop(): Promise<void> {
    if (stopped) return;
    stopped = true;
    await Promise.allSettled([...activeTasks]);
    await plugin.stop();
  }

  const lines = createInterface({ input: stdin, crlfDelay: Infinity });
  for await (const line of lines) {
    if (!line.trim()) continue;
    let message: ProtocolMessage = {};
    try {
      message = JSON.parse(line) as ProtocolMessage;
      if (message.type === "initialize") {
        if (initialized || message.apiVersion !== 2) throw new Error("初始化顺序或 API 版本无效");
        plugin.manifest = message.manifest as Record<string, unknown>;
        plugin.configuration = message.configuration as Record<string, unknown>;
        initialized = true;
        await writer.write({ type: "ready", apiVersion: 2 });
      } else if (message.type === "invoke") {
        if (!initialized) throw new Error("插件尚未初始化");
        parseRequestId(message);
        if (options.concurrentInvocations) {
          const task = invoke(message);
          activeTasks.add(task);
          void task.finally(() => activeTasks.delete(task)).catch((error: unknown) => console.error(error));
        } else {
          await invoke(message);
        }
      } else if (message.type === "stop") {
        await waitAndStop();
        return;
      } else {
        throw new Error("未知 Sidecar 消息");
      }
    } catch (error) {
      const requestId = message.requestId;
      if (typeof requestId !== "number" || !Number.isSafeInteger(requestId) || requestId < 0) {
        console.error(`Sidecar 协议错误：${error instanceof Error ? `${error.name}: ${error.message}` : String(error)}`);
        await waitAndStop();
        return;
      }
      await writer.write({ type: "error", requestId, message: error instanceof Error ? `${error.name}: ${error.message}` : String(error) });
    }
  }
  await waitAndStop();
}

/** 从模块默认导出或 plugin 命名导出载入 Plugin；入口错误时在读循环前失败。 */
async function loadPlugin(entry: string): Promise<Plugin> {
  const module = await import(pathToFileURL(resolve(entry)).href) as { default?: unknown; plugin?: unknown };
  const plugin = module.default ?? module.plugin;
  if (!(plugin instanceof Plugin)) throw new Error("入口必须导出 Plugin 实例");
  return plugin;
}

/** 作为独立 runner 调用时加载作者模块；生产入口也可直接调用 serve(plugin)。 */
async function main(): Promise<void> {
  const argumentsWithoutFlags = process.argv.slice(2).filter((argument) => argument !== "--concurrent");
  const entry = argumentsWithoutFlags[0];
  if (!entry) throw new Error("用法：node runner.js plugin.js");
  await serve(await loadPlugin(entry), { concurrentInvocations: process.argv.includes("--concurrent") });
}

if (process.argv[1] && import.meta.url === pathToFileURL(resolve(process.argv[1])).href) {
  main().catch((error: unknown) => { console.error(error); process.exitCode = 1; });
}

File: Server/PluginSDK/Python/src/capturePluginSdk/runner.py
"""把标准输入上的 Sidecar JSONL 协议映射到普通 Python 插件函数。"""

from __future__ import annotations

import importlib
import json
import sys
import threading

from .model import PROCESS_JSON_SAFE_INTEGER_MAX
from .plugin import Plugin


def loadPlugin(specification: str) -> Plugin:
    """从 module[:attribute] 载入插件；属性可为 Plugin 或返回 Plugin 的无参函数。"""
    moduleName, separator, attributeName = specification.partition(":")
    module = importlib.import_module(moduleName)
    exported = getattr(module, attributeName or "plugin")
    plugin = exported() if callable(exported) and not isinstance(exported, Plugin) else exported
    if not isinstance(plugin, Plugin):
        raise TypeError("入口必须导出 Plugin 或返回 Plugin 的函数")
    return plugin


class JsonLineWriter:
    """串行化多个作者线程的协议输出，保证每个 JSON 帧独占完整一行。"""

    def __init__(self) -> None:
        """创建进程级输出锁；锁只覆盖一次编码和写入，不包围作者函数。"""
        self.lock = threading.Lock()

    def write(self, message: dict[str, object]) -> None:
        """原子写出并刷新单条 JSON；编码或管道失败会传播给当前调度任务。"""
        encoded = json.dumps(message, ensure_ascii=False, separators=(",", ":"))
        with self.lock:
            print(encoded, flush=True)


def run(plugin: Plugin, concurrentInvocations: bool = False) -> None:
    """执行 JSONL；默认串行，作者显式开启后允许 requestId 结果乱序返回。"""
    initialized = False
    writer = JsonLineWriter()
    activeTasks: set[threading.Thread] = set()
    activeTasksLock = threading.Lock()

    def parseRequestId(message: dict[str, object]) -> int:
        """校验进程协议请求号；安全整数上限保证所有官方语言都能无损关联响应。"""
        requestId = message.get("requestId")
        if isinstance(requestId, bool) or not isinstance(requestId, int) or requestId < 0 or requestId > PROCESS_JSON_SAFE_INTEGER_MAX:
            raise ValueError("requestId 必须是非负 JSON 安全整数")
        return requestId

    def invoke(message: dict[str, object]) -> None:
        """执行一个 invoke 并回显 requestId；异常转换为同请求 error 帧。"""
        requestId = parseRequestId(message)
        try:
            action = plugin.invoke(message["invocation"])  # type: ignore[arg-type]
            writer.write({"type": "result", "requestId": requestId, "action": action.toDict()})
        except Exception as error:
            writer.write({"type": "error", "requestId": requestId, "message": f"{type(error).__name__}: {error}"})
        finally:
            with activeTasksLock:
                activeTasks.discard(threading.current_thread())

    def waitAndStop() -> None:
        """等待所有并发作者任务结束后调用一次插件停止生命周期。"""
        while True:
            with activeTasksLock:
                tasks = list(activeTasks)
            if not tasks:
                break
            for task in tasks:
                task.join()
        plugin.stop()

    for line in sys.stdin:
        if not line.strip():
            continue
        message: dict[str, object] = {}
        try:
            message = json.loads(line)
            messageType = message.get("type")
            if messageType == "initialize":
                if initialized or message.get("apiVersion") != 2:
                    raise ValueError("初始化顺序或 API 版本无效")
                plugin.manifest = dict(message.get("manifest", {}))
                plugin.configuration = dict(message.get("configuration", {}))
                initialized = True
                writer.write({"type": "ready", "apiVersion": 2})
                continue
            elif messageType == "invoke":
                if not initialized:
                    raise ValueError("插件尚未初始化")
                parseRequestId(message)
                if concurrentInvocations:
                    task = threading.Thread(target=invoke, args=(message,), name=f"plugin-{message.get('requestId')}")
                    with activeTasksLock:
                        activeTasks.add(task)
                    task.start()
                else:
                    invoke(message)
                continue
            elif messageType == "stop":
                waitAndStop()
                return
            else:
                raise ValueError("未知 Sidecar 消息")
        except Exception as error:
            requestId = message.get("requestId") if isinstance(message, dict) else None
            if isinstance(requestId, bool) or not isinstance(requestId, int) or requestId < 0 or requestId > PROCESS_JSON_SAFE_INTEGER_MAX:
                print(f"Sidecar 协议错误：{type(error).__name__}: {error}", file=sys.stderr)
                waitAndStop()
                return
            response = {"type": "error", "requestId": requestId, "message": f"{type(error).__name__}: {error}"}
        writer.write(response)
    waitAndStop()


def main() -> int:
    """启动独立开发运行器；生产插件也可在自己的入口脚本中直接调用 serve。"""
    arguments = [argument for argument in sys.argv[1:] if argument != "--concurrent"]
    if len(arguments) != 1:
        print("用法：python -m capturePluginSdk.runner module[:attribute] [--concurrent]", file=sys.stderr)
        return 2
    run(loadPlugin(arguments[0]), concurrentInvocations="--concurrent" in sys.argv[1:])
    return 0


if __name__ == "__main__":
    raise SystemExit(main())


def serve(plugin: Plugin, concurrentInvocations: bool = False) -> None:
    """启动 Sidecar 循环；并发模式由作者显式选择，Host 不施加线程数量限制。"""
    run(plugin, concurrentInvocations)
